README.md
# MMM-PrayerTime [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://raw.githubusercontent.com/slametps/MMM-PrayerTime/master/LICENSE)
This an extension for the [MagicMirror<sup>2</sup>](https://github.com/MichMich/MagicMirror). It will display prayer time (in local time) and also recite Adzan/Adhan. Your mirror will be your helpful assistant to do sholat on time.

## Screenshot
![Screenshot](https://raw.githubusercontent.com/slametps/MMM-PrayerTime/master/screenshot.png)
![Screenshot](https://raw.githubusercontent.com/slametps/MMM-PrayerTime/master/screenshot-alert.png)
![Screenshot](https://raw.githubusercontent.com/slametps/MMM-PrayerTime/master/screenshot2.jpg)
![Screenshot](https://raw.githubusercontent.com/slametps/MMM-PrayerTime/master/screenshot-telegram.png)

## Installation
1. Navigate into your MagicMirror's `modules` folder
2. execute `git clone https://github.com/slametps/MMM-PrayerTime.git`
3. if this module does not run correctly, try `npm install async`

## Using the module

To use this module, add it to the modules array in the `config/config.js` file:
````javascript
modules: [
	{
		module: 'MMM-PrayerTime',
		position: 'top_left',	// This can be any of the regions. Best result is in the top_left/top_right.
		config: {
			apiVersion: '1.0', // please, leave unchanged. reserved for future use.
			lat: false, // latitude of your position (city)
			lon: false, // longitude of your position (city)
			timezone: false, // please refer to http://php.net/manual/en/timezones.php
			timeFormat: 24,
			method: 5,
			playAdzan: ['fajr', 'dhuhr', 'asr', 'maghrib', 'isha'],
			notDisplayed: ['midnight', 'sunset'],
			useUpdateInterval: true,
			updateInterval: 86400 * 1000, // How often do you want to fetch new praying time? (milliseconds)
			animationSpeed: 2.5 * 1000, // Speed of the update animation. (milliseconds)
			language: config.language,
			showAdzanAlert: true,
			showTomorrow: true,
			alertTimer: 15000
		}
	}
]
````

## Configuration options

The following properties can be configured:


<table width="100%">
	<!-- why, markdown... -->
	<thead>
		<tr>
			<th>Option</th>
			<th width="100%">Description</th>
		</tr>
	<thead>
	<tbody>
		<tr>
			<td><code>apiVersion</code></td>
			<td>1.0 (reserved for future use)</td>
		</tr>
		<tr>
			<td><code>lat</code><br>REQUIRED</td>
			<td>latitute of your position (city)
				<br>You can use google maps to get this info
				<br><b>Default value:</b> <code>false</code>
			</td>
		</tr>
		<tr>
			<td><code>lon</code><br>REQUIRED</td>
			<td>longitude of your position (city)
				<br>You can use google maps to get this info
				<br><b>Default value:</b> <code>false</code>
			</td>
		</tr>
		<tr>
			<td><code>timezone</code><br>REQUIRED</td>
			<td>Local timezone.
				<br><b>Possible values:</b> refer to http://php.net/manual/en/timezones.php
				<br><b>Default value:</b> <code>true</code>
			</td>
		</tr>
		<tr>
			<td><code>timeFormat</code>
			<td>12/24 time-format.
				<br><b>Possible values:</b> <code>12, 24</code>
				<br><b>Default value:</b> <code>config.timeFormat or 24</code>
			</td>
		</tr>
		<tr>
			<td><code>method</code><br>REQUIRED</td>
			<td>method to identify various schools of thought about how to compute the timings
				<br><b>Possible values:</b> <code>0-7</code> (0-Shia Ithna-Ashari, 1-University of Islamic Sciences,Karachi, 2-Islamic Society of North America (ISNA), 3-Muslim World League/MWL, 4-Umm al-Qura,Makkah, 5-Egyptian General Authority of Survey, 7-Institute of Geophysics, University of Tehran)
				<br><b>Default value:</b> <code>5</code>
			</td>
		</tr>
		<tr>
			<td><code>playAdzan</code></td>
			<td>List/array of the prayer time name to play Adzan.
				<br><b>Possible values:</b> <code>['fajr', 'dhuhr', 'asr', 'maghrib', 'isha', 'imsak', 'sunrise', 'sunset', 'midnight']</code>
				<br><b>Default value:</b> <code>['fajr', 'dhuhr', 'asr', 'maghrib', 'isha']</code>
			</td>
		</tr>
		<tr>
			<td><code>notDisplayed</code></td>
			<td>List/array of the prayer time name not to be displayed.
				<br><b>Possible values:</b> <code>['fajr', 'dhuhr', 'asr', 'maghrib', 'isha', 'imsak', 'sunrise', 'sunset', 'midnight']</code>
				<br><b>Default value:</b> <code>['midnight', 'sunset']</code>
			</td>
		</tr>
		<tr>
			<td><code>useUpdateInterval</code></td>
			<td>Using internal update interval or triggered by MMM-ModuleScheduler to fetch and update new praying time? if <code>false</code>, add new schedule entry in MMM-ModuleScheduler:
				<br><code>{notification: 'PRAYER_TIME', schedule: '1 0 * * *', payload: {type: 'UPDATE_PRAYINGTIME'}}</code>
				<br><b>Possible values:</b> <code>true, false</code>
				<br><b>Default value:</b> <code>true</code>
			</td>
		</tr>
		<tr>
			<td><code>updateInterval</code></td>
			<td>How often to fetch new praying time? (milliseconds)
				<br><b>Default value:</b> <code>86400000</code> (1 day)
			</td>
		</tr>
		<tr>
			<td><code>animationSpeed</code></td>
			<td>Speed of the update animation. (milliseconds)<br>
				<br><b>Possible values:</b><code>0</code> - <code>5000</code>
				<br><b>Default value:</b> <code>2500</code> (2.5 seconds)
			</td>
		</tr>
		<tr>
			<td><code>language</code></td>
			<td>module language. Can be different from MM.
				<br><b>Default value:</b> <code>config.language</code> or <code>'en'</code>
			</td>
		</tr>
		<tr>
			<td><code>showAdzanAlert</code></td>
			<td>Display alert when enterigng praying time?
				<br><b>Possible values:</b> <code>true, false</code>
				<br><b>Default value:</b> <code>true</code>
			</td>
		</tr>
		<tr>
			<td><code>showTomorrow</code></td>
			<td>Display tomorrow praying time?
				<br><b>Possible values:</b> <code>true, false</code>
				<br><b>Default value:</b> <code>true</code>
			</td>
		</tr>
		<tr>
			<td><code>alertTimer</code></td>
			<td>How long alert will be displayed? (milliseconds)<br>
				<br><b>Possible values:</b><code>0</code> - <code>60000</code>
				<br><b>Default value:</b> <code>15000</code> (15 seconds)
			</td>
		</tr>
	</tbody>
</table>

## Dependencies
- Access to the internet to download praying time from http://api.aladhan.com
- [MMM-ModuleScheduler](https://github.com/ianperrin/MMM-ModuleScheduler). OPTIONAL.
- omxplayer (<code>sudo apt-get install omxplayer</code>)
- [MMM-TelegramBot](https://github.com/eouia/MMM-TelegramBot). OPTIONAL.

The MIT License (MIT)
=====================

Copyright © 2016-2017 Slamet PS

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the “Software”), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

**The software is provided “as is”, without warranty of any kind, express or implied, including but not limited to the warranties of merchantability, fitness for a particular purpose and noninfringement. In no event shall the authors or copyright holders be liable for any claim, damages or other liability, whether in an action of contract, tort or otherwise, arising from, out of or in connection with the software or the use or other dealings in the software.**


CHANGELOG.md
# MMM-PrayerTime Change Log
All notable changes to this project will be documented in this file.

## [1.0.2] - 2019-04-16

### Added
- option to display tomorrow praying time or not (default = true)

## [1.0.1] - 2018-05-17

### Changed
- 'en' as default language if language option is not defined nor MM language option

### Added
- add 12-hours time-format support
- view praying-time in telegram via command ````/prayertime````. requires [MMM-TelegramBot](https://github.com/eouia/MMM-TelegramBot) to work.
- custom-sound for IMSAK

## [1.0.0] - 2017-04-01
** INITIAL RELEASE**

### Added
- today and tomorrow praying-time
- translations support (id, en)

node_helper.js
/* Magic Mirror
 * Node Helper: MMM-PrayerTime
 *
 * By Slamet PS/slametps@gmail.com
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
var async = require('async');
var exec = require('child_process').exec;

module.exports = NodeHelper.create({
	// Subclass start method.
	start: function() {
		console.log("Starting node_helper.js for MMM-PrayerTime.");
	},

	socketNotificationReceived: function(notification, payload) {
    console.log(this.name + " node helper received a socket notification: " + notification + " - Payload: " + payload);
    if (notification == "PLAY_ADZAN") {
      var adzanSound = 'adzan.mp3';
      if (payload.occasion) {
        if (payload.occasion=="FAJR") {
          adzanSound = 'adzan-fajr.mp3';
        }
        else if (payload.occasion=="IMSAK") {
          adzanSound = 'imsak.mp3';
        }
      }
      var adzanCmd = '/usr/bin/omxplayer  modules/MMM-PrayerTime/res/' + adzanSound + ' &';
      async.parallel([
        async.apply(exec, adzanCmd)
      ],
      function (err, res) {
      });
		}
	},
});


MMM-PrayerTime.js
Module.register("MMM-PrayerTime",{
	// Default module config.
	defaults: {
		apiVersion: '1.0',
    lat: false,
    lon: false,
    timezone: false,
    timeFormat: config.timeFormat || 24,
    method: 5, // method of timing computation {0-Shia Ithna-Ashari,1-University of Islamic Sciences, Karachi,2-Islamic Society of North America (ISNA),3-Muslim World League (MWL),4-Umm al-Qura, Makkah,5-Egyptian General Authority of Survey,7-Institute of Geophysics, University of Tehran}
    playAdzan: ['fajr', 'dhuhr', 'asr', 'maghrib', 'isha'],
    notDisplayed: ['midnight', 'sunset'],
    useUpdateInterval: true,
    updateInterval: 86400 * 1000, // How often do you want to fetch new praying time? (milliseconds)
    animationSpeed: 2.5 * 1000, // Speed of the update animation. (milliseconds)
    language: config.language || "en",
    colored: false,
    showAdzanAlert: true,
    showTomorrow: true,
    alertTimer: 15000
	},

	getScripts: function() {
	    return ["moment.js"];
	},

	getStyles: function() {
		return ["MMM-PrayerTime.css"];
	},

  // Define required translations.
	getTranslations: function() {
    return {
      'en': 'translations/en.json',
      'id': 'translations/id.json',
      'ar': 'translations/ar.json'
    };
	},

  getCommands: function(commander) {
    commander.add({
      command: 'prayertime',
      description: this.translate("TXT_PRAYERTIME_DESC"),
      callback: 'cmd_prayertime'
    })
  },

  cmd_prayertime: function(command, handler) {
    var text = "";
    text += "*" + this.translate("TXT_PRAYERTIME") + "*\n";
    text += "*" + this.translate("TODAY") + "*\n";
    for (var t in this.arrTodaySchedule) {
      text += "*" + this.translate(this.arrTodaySchedule[t][0].toUpperCase()) + ":* `" + (this.config.timeFormat == 12 ? moment(this.arrTodaySchedule[t][1], ["HH:mm"]).format("h:mm A") : this.arrTodaySchedule[t][1]) + "`\n";
    }
    text += "\n*" + this.translate("TOMORROW") + "*\n";
    for (var t in this.arrNextdaySchedule) {
      text += "*" + this.translate(this.arrNextdaySchedule[t][0].toUpperCase()) + ":* `" + (this.config.timeFormat == 12 ? moment(this.arrNextdaySchedule[t][1], ["HH:mm"]).format("h:mm A") : this.arrNextdaySchedule[t][1]) + "`\n";
    }
    handler.reply("TEXT", text, {parse_mode:'Markdown'});
  },


  /* getParams
	 * Generates an url with api parameters based on the config.
	 *
	 * return String - URL params.
	 */
	getParams: function(unixTime) {
		var params = unixTime + "?";
		if(this.config.lat) {
			params += "latitude=" + this.config.lat;
		}
    if (this.config.lon) {
			params += "&longitude=" + this.config.lon;
		}
    if (this.config.timezone) {
			params += "&timezonestring=" + this.config.timezone;
    }
		if (this.config.method) {
			params += "&method=" + this.config.method;
    }

		return params;
	},

	/* processSchedule
	 * process downloaded scheduled.
	 */
  processSchedule: function() {
    var self = this;

    function sortSchedule(a, b) {
      if (a[1] < b[1]) {
        return -1;
      }
      if (a[1] > b[1]) {
        return 1;
      }

      // names must be equal
      return 0;
    }

    // sort today schedule
    this.arrTodaySchedule = [];
    this.arrAdzanTime = [];
    for(var x in this.todaySchedule){
      if (!self.config.notDisplayed.includes(x.toLowerCase()))
        this.arrTodaySchedule.push([x, this.todaySchedule[x]]);
      if (self.config.playAdzan.includes(x.toLowerCase()))
        this.arrAdzanTime.push(this.todaySchedule[x]);
    }
    this.arrTodaySchedule.sort(sortSchedule);

    // sort nextday schedule
    this.arrNextdaySchedule = [];
    for(var x in this.nextdaySchedule){
      if (!self.config.notDisplayed.includes(x.toLowerCase()))
        this.arrNextdaySchedule.push([x, this.nextdaySchedule[x]]);
    }
    this.arrNextdaySchedule.sort(sortSchedule);

    this.loaded = true;
		this.updateDom(this.config.animationSpeed);
  },

  updateSchedule: function(delay) {
    var self = this;
    Log.log(self.name + ': updateSchedule');
    var urlBase = "http://api.aladhan.com/timings/";
    var curUnixTime = moment().unix();
    var urlToday = urlBase + this.getParams(curUnixTime);
    var urlNextday = urlBase + this.getParams(curUnixTime + 86400);
    var resultToday = {};
    var resultNextday = {};
    var nbReq = 2;
    var nbRes = 0;

    var todayRequest = new XMLHttpRequest();
		todayRequest.open("GET", urlToday, true);
		todayRequest.onreadystatechange = function() {
			if (this.readyState === 4) {
				if (this.status === 200) {
          resultToday = JSON.parse(this.responseText);
          self.todaySchedule = resultToday.data.timings;
          // debug/testing only
          //self.todaySchedule = {"Fajr":"04:30", "Dhuhr":"12:00", "Asr":"16:14", "Maghrib":"18:00", "Isha":"20:50", "Imsak":"04:20"};
          nbRes++;
          if (nbRes == nbReq)
            self.processSchedule();
				} else {
					Log.error(self.name + ": got HTTP status-" + this.status);
          retry = true;
				}
			}
		};
		todayRequest.send();

    var nextdayRequest = new XMLHttpRequest();
		nextdayRequest.open("GET", urlNextday, true);
		nextdayRequest.onreadystatechange = function() {
			if (this.readyState === 4) {
				if (this.status === 200) {
          resultNextday = JSON.parse(this.responseText);
          self.nextdaySchedule = resultNextday.data.timings;
          nbRes++;
          if (nbRes == nbReq)
            self.processSchedule();
				} else {
					Log.error(self.name + ": got HTTP status-" + this.status);
          retry = true;
				}
			}
		};
		nextdayRequest.send();
  },

  isAdzanNow: function() {
    var curTime = moment().format("HH:mm:ss");
    var indexAdzan = -1;
    //console.log(this.arrTodaySchedule);
    if (this.arrTodaySchedule.length > 0)
    {
      function isAdzan(el, idx, arr) {
        return (el[1] + ':00') == curTime;
      }
      indexAdzan = this.arrTodaySchedule.findIndex(isAdzan);
      //console.log("indexAdzan-"+indexAdzan);

      if (indexAdzan > -1) {
        //console.log(this.config.playAdzan);
        //console.log("this.arrTodaySchedule[indexAdzan][0]).toLowerCase()-"+(this.arrTodaySchedule[indexAdzan][0]).toLowerCase());
        //console.log("this.config.playAdzan.findIndex((this.arrTodaySchedule[indexAdzan][0]).toLowerCase())-"+this.config.playAdzan.findIndex((this.arrTodaySchedule[indexAdzan][0]).toLowerCase()));
        if (this.config.playAdzan.includes((this.arrTodaySchedule[indexAdzan][0]).toLowerCase())) {
          if (this.config.showAdzanAlert) {
            var occasionNameUpper = (this.arrTodaySchedule[indexAdzan][0]).toUpperCase();
            var alertMsg = "ALERT_ADZAN_MSG";
            var adzanImsak = "ADZAN";
            //console.log("occasionNameUpper-"+occasionNameUpper);
            if (occasionNameUpper == "IMSAK") {
              alertMsg = "ALERT_IMSAK_MSG";
              adzanImsak = "IMSAK";
            }
            this.sendNotification("SHOW_ALERT", {title: this.translate(adzanImsak).toUpperCase(), imageFA: 'bullhorn', message: this.translate(alertMsg).replace("%OCCASION", this.translate(occasionNameUpper)), timer: this.config.alertTimer});
          }
          //console.log("this.arrTodaySchedule[indexAdzan][0]).toUpperCase()-"+(this.arrTodaySchedule[indexAdzan][0]).toUpperCase());
          this.sendSocketNotification("PLAY_ADZAN", {occasion: (this.arrTodaySchedule[indexAdzan][0]).toUpperCase()});
        }
      }
    }
  },

	start: function() {
		Log.info("Starting module: " + this.name);
		var self = this;

    // Set locale.
		moment.locale(this.config.language);

    this.todaySchedule = {};
    this.nextdaySchedule = {};
    this.arrTodaySchedule = [];
    this.arrNextdaySchedule = [];
    this.arrAdzanTime = [];

    this.loaded = false;
    var self = this;

    // first update
    self.updateSchedule(0);
    // periodic update if defined
    if (self.config.useUpdateInterval) {
      Log.log(self.name + ': using periodic update is activated');
      setInterval(function() {
        self.updateSchedule(0);
      }, self.config.updateInterval);
    }
    // adzan-checker
    self.isAdzanNow();
    setInterval(function() {
      self.isAdzanNow();
    }, 1000);
	},

	// Override dom generator.
	getDom: function() {
		Log.log("Updating MMM-PrayerTime DOM.");
    var self = this;
    var wrapper = document.createElement("div");

    if (!this.loaded) {
			wrapper.innerHTML = this.translate("LOADING");
			wrapper.className = "dimmed light small";
		}
    else {
      var table = document.createElement("table");
		  table.className = "small";

      var row = document.createElement("tr");
      if (this.config.colored) {
        row.className = "colored";
      }
      table.appendChild(row);

      var occasionName = document.createElement("td");
      occasionName.className = "occasion-name bright light";
      occasionName.innerHTML = '&nbsp;';
      row.appendChild(occasionName);

      // today
      var occasionTime = document.createElement("td");
      occasionTime.className = "occasion-time bright light";
      occasionTime.innerHTML = this.translate('TODAY');
      row.appendChild(occasionTime);

      if (this.config.showTomorrow) {
        // nextday
        var occasionTimeNext = document.createElement("td");
        occasionTimeNext.className = "occasion-time bright light";
        //occasionTimeNext.innerHTML = this.todaySchedule[t];
        occasionTimeNext.innerHTML = this.translate('TOMORROW');
        row.appendChild(occasionTimeNext);
      }

      //for (var i = 0, count = this.todaySchedule.length; i < count; i++) {
      //for (t in this.todaySchedule)
      for (t in this.arrTodaySchedule)
      {
        row = document.createElement("tr");
        if (this.config.colored) {
          row.className = "colored";
        }
        table.appendChild(row);

        var occasionName = document.createElement("td");
        occasionName.className = "occasion-name bright light";
        //occasionName.innerHTML = this.translate(t);
        occasionName.innerHTML = this.translate(this.arrTodaySchedule[t][0].toUpperCase());
        row.appendChild(occasionName);

        // today
        var occasionTime = document.createElement("td");
        occasionTime.className = "occasion-time bright light";
        //occasionTime.innerHTML = this.todaySchedule[t];
        occasionTime.innerHTML = (this.config.timeFormat == 12 ? moment(this.arrTodaySchedule[t][1], ["HH:mm"]).format("h:mm A") : this.arrTodaySchedule[t][1]);
        row.appendChild(occasionTime);

        if (this.config.showTomorrow) {
          // nextday
          var occasionTimeNext = document.createElement("td");
          occasionTimeNext.className = "occasion-time bright light";
          //occasionTimeNext.innerHTML = this.todaySchedule[t];
          occasionTimeNext.innerHTML = (this.config.timeFormat == 12 ? moment(this.arrNextdaySchedule[t][1], ["HH:mm"]).format("h:mm A") : this.arrNextdaySchedule[t][1]);
          row.appendChild(occasionTimeNext);
        }
      }

      wrapper.appendChild(table);
    }

		return wrapper;
  },

	notificationReceived: function(notification, payload, sender) {
		Log.log(this.name + ": received notification : " + notification);
		if (notification == "PRAYER_TIME") {
      if (payload.type == "PLAY_ADZAN") {
        if (this.config.showAdzanAlert)
          this.sendNotification("SHOW_ALERT", {title: this.translate("ADZAN"), message: this.translate("ALERT_ADZAN_MSG").replace("%OCCASION", this.translate("ASR")), timer: this.config.alertTimer});
        this.sendSocketNotification("PLAY_ADZAN", {occasion: 'ASR'});
      }
      if (payload.type == "UPDATE_PRAYINGTIME") {
        this.updateSchedule(0);
      }
		}
	}
});
